doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,144 @@
+
+* Adding metrics gets more difficult. You not only have to figure out how go get your number counted the right way, you also have to figure out how to aggregate it properly in the dataflow world. We worry that this might mean that useful operational metrics don't get added as soon/as quickly as they'd be necessary, making it harder to run the system.
+
+### Do not scrape arbitrary prometheus endpoints, instead scrape our internal metrics registry
+
+This would get around the need to introduce an externally-definable source, and automatically feed a dataflow that we can keep up to date automatically.
+
+There's one major reason why I think this would be harder to implement right now: Being able to pull in all our metrics depends on [#5825], which would be a pretty heavy re-work of our internal structures (have to thread a registry through to various bits that don't have a place to hold that registry yet, like the pgwire implementation).
+
+Besides that, I think it's somewhat nice to represent this data as a source that we poll, and internally, our metrics could still be realized like this once a fix for [#5825] lands.
+
+[#5825]: https://github.com/MaterializeInc/materialize/issues/5825
+
+## Alternative details
+
+### Table schema - denormalization
+
+Several ways to represent the metrics imported. The most important "meta" information that users will need is the [metric type](https://prometheus.io/docs/concepts/metric_types/) - gauges, counters and others all behave differently, after all.
+
+It would make sense (but be somewhat denormalized) to list the metric type alongside each reading, like so:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    metric_type TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+```
+
+With this schema in place we could leave out the "meta" schema for the first iteration, as it provides only the "help" text.
+
+Another way to avoid the `metric_type` entirely is to use a value column per metric type (and fill only the column whose metric type corresponds to the column):
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+
+    -- These are nice and easy:
+    gauge_value float8,
+    counter_value float8,
+
+    -- And then histograms is where it gets gross:
+    histogram_sum float8,
+    histogram_p50 float8,
+    histogram_p90 float8,
+    -- (...and others, configured in a WITH parameter)
+    histogram_count float8
+);
+```
+
+I think if we want to go down this road, we'll want something like an "is a" table relationship instead of all those NULLable columns. It does smell much more complex and harder to use correctly, though.
+
+# Open questions
+
+* Should we expose the CREATE SOURCE mechanics for this externally? It probably
```

Comment:
I think the generic scraper would be awesome, particularly for being able to write quality alerts from Prometheus metrics that need to join data against external sources.